README.md
# Typography test page for Silverstripe

A simple extension (extending Page) to add a simple typography test page to see your website styles.

It includes a JavaScript function to add element titles to all elements in the test page to help you see what element is which.


## Requirements

- Silverstripe ^4.0 || ^5.0 in `dev` environment

For Silverstripe 3, please refer to the [Silverstripe3 branch](https://github.com/axllent/silverstripe-typography/tree/silverstripe3).


## Installation

```
composer require axllent/silverstripe-typography --dev
```


## Usage

Install the module, run a `?flush=1` and access your website with /typo/ (eg: www.example.com/typo/).
Please note that your website must be in the `dev` environment for the route to be activated.

If you need a customised layout then add a Layout/Typography.ss template in your theme directory and

```
    <% include TypographySampleText %>
```


CHANGELOG.md
# Changelog

Notable changes to this project will be documented in this file.

## [2.0.4]

- Support for Silverstripe 5


## [2.0.3]

- Make `index()` compatible with `PageController::index(SilverStripe\Control\HTTPRequest $request)`


## [2.0.2]

- Switch to silverstripe-vendormodule


## [2.0.1]

- Remove CountryDropDownField
- Add FieldGroup


## [2.0.0]

- SilverStripe 4 support


## [1.1.1]

- Fix SilverStripe framework version requirement


## [1.1.0]

- Switch to using SilverStripe routes to become compatible with SS 3.3
- Rename template & include file
- Allow only in dev environment
- Compatible with SS >=3


## [1.0.0]

- Adopt semantic versioning releases


javascript/typo.js
function tagsToTitle() {
	var els = document.getElementById("TypoTestPage").getElementsByTagName("*");
	for (var i = 0, max = els.length; i < max; i++) {
		var el = els[i];
		var classes = [];

		var t = el.tagName;
		var s = el.getAttribute('style');
		var c = el.getAttribute('class');
		if (s) t = t + '[style="' + s + '"]';
		else if (c) t = t + '[class="' + c + '"]';
		classes.push(t);

		var p = el.parentNode;
		while (p && p.id != "TypoTestPage") {
			var t = p.tagName;
			var s = p.getAttribute('style');
			var c = p.getAttribute('class');
			if (s) t = t + '[' + s + ']';
			else if (c) t = t + '[class="' + c + '"]';
			classes.push(t);
			p = p.parentNode;
		}
		classes.reverse();
		els[i].title = classes.join(' > ').toLowerCase();
	}
}

var w = window;
w.addEventListener ? w.addEventListener("load", tagsToTitle, !1) : w.attachEvent && w.attachEvent("onload", tagsToTitle);


code/Typography.php
<?php
/**
 * Typography test page for SilverStripe
 * =====================================
 *
 * A SilerStripe extension to add a typography test page to your website
 *
 * Once installed, run a flush=1 and access /typo/ on your website
 * eg: www.exmaple.com/typo/
 *
 * License: MIT-style license http://opensource.org/licenses/MIT
 * Author: Techno Joy development team (www.technojoy.co.nz)
 * Inspired by https://github.com/sunnysideup/silverstripe-typography
 */
class Typo extends Page_Controller
{

    public function init()
    {
        parent::init();
        Requirements::javascript(
            basename(dirname(dirname(__FILE__))) . '/javascript/typo.js'
        );
        $this->Title = 'Typography test page';
        $this->ExtraMeta .= '<meta name="robots" content="noindex, nofollow" />';
    }

    function TypoForm()
    {
        $array= array('green', 'yellow', 'blue', 'pink', 'orange');
        $form = new Form(
            $this,
            'TestForm',
            $fields = FieldList::create(
                HeaderField::create('HeaderField1', 'HeaderField Level 1', 1),
                LiteralField::create('LiteralField', '<p>All fields up to EmailField are required and should be marked as such</p>'),
                TextField::create('TextField1', 'Text Field Example 1'),
                TextField::create('TextField2', 'Text Field Example 2'),
                TextField::create('TextField3', 'Text Field Example 3'),
                TextField::create('TextField4', ''),
                HeaderField::create('HeaderField2b', 'Field with right title', 2),
                $textAreaField = new TextareaField('TextareaField', 'Textarea Field'),
                EmailField::create('EmailField', 'Email address'),
                HeaderField::create('HeaderField2c', 'HeaderField Level 2', 2),
                DropdownField::create('DropdownField', 'Dropdown Field', array( 0 => '-- please select --', 1 => 'test AAAA', 2 => 'test BBBB')),
                OptionsetField::create('OptionSF', 'Optionset Field', $array),
                CheckboxSetField::create('CheckboxSF', 'Checkbox Set Field', $array),
                CountryDropdownField::create('CountryDropdownField', 'Countries'),
                CurrencyField::create('CurrencyField', 'Bling bling', '$123.45'),
                HeaderField::create('HeaderField3', 'Other Fields', 3),
                NumericField::create('NumericField', 'Numeric Field '),
                DateField::create('DateField', 'Date Field'),
                DateField::create('DateTimeField', 'Date and Time Field'),
                CheckboxField::create('CheckboxField', 'Checkbox Field')
            ),
            $actions = FieldList::create(
                FormAction::create('submit', 'Submit Button')
            ),
            $requiredFields = RequiredFields::create(
                'TextField1',
                'TextField2',
                'TextField3',
                'ErrorField1',
                'ErrorField2',
                'EmailField',
                'TextField3',
                'RightTitleField',
                'CheckboxField',
                'CheckboxSetField'
            )
        );
        $textAreaField->setColumns(45);
        $form->setMessage('warning message', 'warning');
        return $form;
    }

    function TestForm($data)
    {
        $this->redirectBack();
    }
}


code/Typography_Controller.php
<?php
/**
 * Typography test page for SilverStripe
 * =====================================
 *
 * A SilerStripe extension to add a typography test page to your website
 *
 * Once installed, run a flush=1 and access /typo/ on your website
 * eg: www.exmaple.com/typo/
 *
 * License: MIT-style license http://opensource.org/licenses/MIT
 * Author: Techno Joy development team (www.technojoy.co.nz)
 * Inspired by https://github.com/sunnysideup/silverstripe-typography
 */
class Typography_Controller extends Page_Controller
{

    public function index()
    {
        Requirements::javascript(
            basename(dirname(dirname(__FILE__))) . '/javascript/typo.js'
        );
        $this->Title = 'Typography test page';
        $this->ExtraMeta .= '<meta name="robots" content="noindex, nofollow" />';

        return $this->renderWith(array('Typography', 'Page'));
    }

    function TypoForm()
    {
        $array= array('green', 'yellow', 'blue', 'pink', 'orange');
        $form = new Form(
            $this,
            'TestForm',
            $fields = FieldList::create(
                HeaderField::create('HeaderField1', 'HeaderField Level 1', 1),
                LiteralField::create('LiteralField', '<p>All fields up to EmailField are required and should be marked as such</p>'),
                TextField::create('TextField1', 'Text Field Example 1'),
                TextField::create('TextField2', 'Text Field Example 2'),
                TextField::create('TextField3', 'Text Field Example 3'),
                TextField::create('TextField4', ''),
                HeaderField::create('HeaderField2b', 'Field with right title', 2),
                $textAreaField = new TextareaField('TextareaField', 'Textarea Field'),
                EmailField::create('EmailField', 'Email address'),
                HeaderField::create('HeaderField2c', 'HeaderField Level 2', 2),
                DropdownField::create('DropdownField', 'Dropdown Field', array( 0 => '-- please select --', 1 => 'test AAAA', 2 => 'test BBBB')),
                OptionsetField::create('OptionSF', 'Optionset Field', $array),
                CheckboxSetField::create('CheckboxSF', 'Checkbox Set Field', $array),
                CountryDropdownField::create('CountryDropdownField', 'Countries'),
                CurrencyField::create('CurrencyField', 'Bling bling', '$123.45'),
                HeaderField::create('HeaderField3', 'Other Fields', 3),
                NumericField::create('NumericField', 'Numeric Field '),
                DateField::create('DateField', 'Date Field'),
                DateField::create('DateTimeField', 'Date and Time Field'),
                CheckboxField::create('CheckboxField', 'Checkbox Field')
            ),
            $actions = FieldList::create(
                FormAction::create('submit', 'Submit Button')
            ),
            $requiredFields = RequiredFields::create(
                'TextField1',
                'TextField2',
                'TextField3',
                'ErrorField1',
                'ErrorField2',
                'EmailField',
                'TextField3',
                'RightTitleField',
                'CheckboxField',
                'CheckboxSetField'
            )
        );
        $textAreaField->setColumns(45);
        $form->setMessage('warning message', 'warning');
        return $form;
    }

    function TestForm($data)
    {
        $this->redirectBack();
    }
}


src/TypographyController.php
<?php

namespace Axllent\Typography;

use SilverStripe\Control\Director;
use SilverStripe\Forms\CheckboxField;
use SilverStripe\Forms\CheckboxSetField;
use SilverStripe\Forms\CurrencyField;
use SilverStripe\Forms\DateField;
use SilverStripe\Forms\DateTimeField;
use SilverStripe\Forms\DropdownField;
use SilverStripe\Forms\EmailField;
use SilverStripe\Forms\FieldGroup;
use SilverStripe\Forms\FieldList;
use SilverStripe\Forms\Form;
use SilverStripe\Forms\FormAction;
use SilverStripe\Forms\HeaderField;
use SilverStripe\Forms\LiteralField;
use SilverStripe\Forms\NumericField;
use SilverStripe\Forms\OptionsetField;
use SilverStripe\Forms\RequiredFields;
use SilverStripe\Forms\TextareaField;
use SilverStripe\Forms\TextField;
use SilverStripe\View\Requirements;

/**
 * Typography test page for SilverStripe
 * =====================================
 *
 * A Silverstripe extension to add a typography test page to your website
 *
 * Once installed, run a flush=1 and access /typo/ on your website
 * eg: www.example.com/typo/
 *
 * License: MIT-style license http://opensource.org/licenses/MIT
 * Author: Techno Joy development team (www.technojoy.co.nz)
 * Inspired by https://github.com/sunnysideup/silverstripe-typography
 */
class TypographyController extends \PageController
{
    /**
     * Allowed actions
     *
     * @var array
     *
     * @config
     */
    private static $allowed_actions = [
        'TestForm',
    ];

    public function link($action = null)
    {
        return Director::baseURL() . 'typo';
    }

    /**
     * Index
     *
     * @param HTTPRequest $request HTTP request
     *
     * @return HTTPResponse
     */
    public function index($request)
    {
        Requirements::javascript('axllent/silverstripe-typography: javascript/typography.js');

        $this->Title = 'Typography test page';
        $this->ExtraMeta .= '<meta name="robots" content="noindex, nofollow">';

        return $this->renderWith(['Typography', 'Page']);
    }

    /**
     * Typo form
     *
     * @return Form
     */
    public function typoForm()
    {
        $array = ['green', 'yellow', 'blue', 'pink', 'orange'];
        $form  = Form::create(
            $this,
            'TestForm',
            $fields = FieldList::create(
                HeaderField::create('HeaderField1', 'HeaderField Level 1', 1),
                LiteralField::create('LiteralField', '<p>All fields up to EmailField are required and should be marked as such</p>'),
                TextField::create('TextField1', 'Text Field Example 1'),
                TextField::create('TextField2', 'Text Field Example 2'),
                TextField::create('TextField3', 'Text Field Example 3'),
                TextField::create('TextField4', ''),
                HeaderField::create('FieldGroupHdr', 'First/last name FieldGroup'),
                FieldGroup::create(
                    TextField::create('FirstName1', 'First Name'),
                    TextField::create('LastName1', 'Last Name')
                ),
                HeaderField::create('HeaderField2b', 'Field with right title', 2),
                TextareaField::create('TextareaField', 'Textarea Field')
                    ->setColumns(45)
                    ->setRightTitle('This is the right title'),
                EmailField::create('EmailField', 'Email address'),
                HeaderField::create('HeaderField2c', 'HeaderField Level 2', 2),
                DropdownField::create(
                    'DropdownField',
                    'Dropdown Field',
                    [
                        0 => '-- please select --',
                        1 => 'test AAAA',
                        2 => 'test BBBB'
                    ]
                ),
                OptionsetField::create('OptionSF', 'Optionset Field', $array),
                CheckboxSetField::create('CheckboxSF', 'Checkbox Set Field', $array),
                CurrencyField::create('CurrencyField', 'Bling bling', '$123.45'),
                HeaderField::create('HeaderField3', 'Other Fields', 3),
                NumericField::create('NumericField', 'Numeric Field '),
                DateField::create('DateField', 'Date Field'),
                DateTimeField::create('DateTimeField', 'Date and Time Field'),
                CheckboxField::create('CheckboxField', 'Checkbox Field')
            ),
            $actions = FieldList::create(
                FormAction::create('submit', 'Submit Button')
            ),
            $requiredFields = RequiredFields::create(
                'TextField1',
                'TextField2',
                'TextField3',
                'ErrorField1',
                'ErrorField2',
                'EmailField',
                'TextField3',
                'RightTitleField',
                'CheckboxField',
                'CheckboxSetField'
            )
        );

        $form->setMessage('warning message', 'warning');

        return $form;
    }

    /**
     * Simple redirect function
     *
     * @param array $data Form data
     *
     * @return HTTPResponse
     */
    public function testForm($data)
    {
        return $this->redirectBack();
    }
}
